File: leetcode/100_199/158__Read_N_Characters_Given_Read4_2.java
/**
 * The read4 API is defined in the parent class Reader4.
 *     int read4(char[] buf); 
 */
public class Solution extends Reader4 {
    private char[] tmp = new char[4];
    private int readIdx = 0;
    private int size = 0;
    
    /**
     * @param buf Destination buffer
     * @param n   Number of characters to read
     * @return    The number of actual characters read
     */
    public int read(char[] buf, int n) {
        if (buf == null || buf.length == 0 || n <= 0) {
            return 0;
        }
        if (n > buf.length) {
            n = buf.length;
        }
        
        int index = 0;
        while(index < n) {
            if (readIdx == size) {
                size = read4(tmp);
                readIdx = 0;
                if (size == 0) {
                    break;
                }
            }
            buf[index++] = tmp[readIdx++];
        }
        return index;
    }
}

File: leetcode/100_199/158__Read_N_Characters_Given_Read4_2.cpp
// Forward declaration of the read4 API.
int read4(char *buf);

class Solution {
private:
    char tmp[4];
    int readIdx = 0;
    int tmpSize = 0;

public:
    /**
     * @param buf Destination buffer
     * @param n   Number of characters to read
     * @return    The number of actual characters read
     */
    int read(char *buf, int n) {
        int index = 0;
        while(index < n) {
            if (readIdx == tmpSize) {
                tmpSize = read4(tmp);
                readIdx = 0;
            }
            for (; index < n && readIdx < tmpSize; index++, readIdx++) {
                buf[index] = tmp[readIdx];
            }
            if (tmpSize < 4) {
                break;
            }
        }
        return index;
    }
};

// Forward declaration of the read4 API.
int read4(char *buf);

class Solution {
private:
    char tmp[4];
    int readIdx = 0;
    int tmpSize = 0;

public:
    /**
     * @param buf Destination buffer
     * @param n   Number of characters to read
     * @return    The number of actual characters read
     */
    int read(char *buf, int n) {
        int index = 0;
        while(index < n) {
            if (tmpSize == 0) {
                tmpSize = read4(tmp);
                readIdx = 0;
                if (tmpSize == 0) {
                    break;
                }
            }
            buf[index++] = tmp[readIdx++];
            tmpSize--;
        }
        return index;
    }
};